articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+
+<!--
+1. Pass data to the caller
+2. Transfer control to callee
+3. Save the state of the caller
+4. Run callee code
+5. Set up the return value
+6. Restore saved state of the caller
+7. Transfer control back to the caller
+8. Finish caller code
+-->
+
+When we call a function we want our ABI defined process to be dynamic. This means that we want to be able to do all the steps correctly given that the only things that the caller will know are the signature and where the callee is defined in memory. And the only thing the callee will know is its own signature. We also want to be able to have theoretically infinite function calling depth (memory bounded <!-- better word please -->of course).
+
+So how does this get done? Well, let's look at calling the function first. 
+
+## The Stack
+
+The stack is how we will be able to keep track of who called the callee, more specifically were in the caller the callee needs to return to in memory. This is called the return address. The stack will also help with saving and resorting the caller's state so then when the callee returns to the caller, the caller can pick up where it left off.
+
+The stack works in the way you would expect from the name, i.e. we can push things on to the top of the stack and we can pop the top element from the stack. I am grossly oversimplifying the functionality of the stack but this article is not meant to talk about all of the ins and outs of the stack.
+
+### Transferring Control
+
+First a little background on how the CPU executed a list of instructions. To know which instruction is needing to be executed there is a register called the instruction pointer, `rip`, that points to the next instruction needing to be executed. This will step through the code one instruction at a time.
+
+Back to the topic, in assembly code, we would call a function with the `call <callee addr>` instruction. This will do two things; First, it will push the current value of the instruction pointer on to the stack (this is the return address) and then it will change the value of the instruction pointer to be the address of the callee function.
+
+When we want to return from the callee and go back to the caller we can use the assembly instruction `ret` which will basically do the opposite of what `call` did. It will pop the return address off of the stack and into `rip` which will set the next instruction (what `rip` is pointing at) to be the instruction after the `call` in the caller code.
+
+Now you can think about how we could just keep calling function which would just keep pushing return addresses onto the stack until we choose to return which would pop all of the return addresses one at a time off of the stack until we are back to where we started.
+
+<!-- animation maybe, I don't really want to make one -->
+
+Calling and return from functions is important but it's more of a hardware thing then it as an ABI thing. So let's see how the ABI build off of this to allows for argument passing.

```

Comment:
it's more of a hardware thing then it as an ABI thing. -> it  is normally handled by hardware rather than the ABI.

Reply:
So let's see -> Let's see